### ISSUE TITLE
pure component lib for choo & friends

### ISSUE THREAD

**Issue by ungoldman** [ungoldman/ideas#15]
nanocomponent but dumber & with standardized function signature

<img width="754" alt="screen shot 2017-08-15 at 2 26 52 pm" src="https://user-images.githubusercontent.com/427322/29337380-ddf2fee6-81c5-11e7-93b4-73b61a22cea6.png">

cc @bcomnes

**Comment by bcomnes** [comment]
I like some of the ideas here because they might enable some interesting things around automatic instance management. You should implement it!

**Comment by yoshuawuyts** [comment]
I'd do `component(name, options, children)` probably. The name argument makes for heaps good debugging / tracing

**Comment by yoshuawuyts** [comment]
Oh, and if every component is an emitter, then having a `.use()` method might simply be:
```js
var button = buttonComponent()

// since there's no "state" argument, it's just "emitter", and "this"
button.use(function (emitter, button) {
  emitter.on('*', function (eventName, data) {
    console.log(button.name + ':' + eventName, data)
  })
})
```
woot!

**Comment by ungoldman** [comment]
name for tracing is a good idea! I'd be partial to having it as an option in the first object to keep it consistent with the overall idea of [h2spec](https://github.com/hyper2/h2spec) tho

**Comment by ungoldman** [comment]
For the emitter concept I was thinking the other way around, where button emits events up, and any state changes are just passed down in a new options object. (["data down, actions up."](https://github.com/maxogden/yo-yo)).

`use`-style middleware for every component seems like it might become quickly unmanageable

**Comment by ungoldman** [comment]
@bcomnes mentioned with some apprehension that this sounded like it was approaching a redux-like `connect` pattern -- wondering if there's another smart/simple way to have components emit actions up to choo's app state.

**Comment by ungoldman** [comment]
![fullsizerender](https://user-images.githubusercontent.com/427322/29385824-ffd9884a-828d-11e7-8dd3-c7087ab504c6.jpg)

**Comment by ungoldman** [comment]
maybe the solution is to pass choo's `emit` arg when instantiating a component in a view's constructor? though views !== components in choo

**Comment by timwis** [comment]
Interested to see where this goes! Hard to imagine it just from the doctors note attached to the issue (tee-hee) but looking forward to hearing more!

**Comment by yoshuawuyts** [comment]
Passing the emitter + state to the component constructor is pretty easy
with https://github.com/yoshuawuyts/shared-component/

**Comment by ungoldman** [comment]
I've made a very simple library called [hyperaxe](https://github.com/ungoldman/hyperaxe) that wraps hyperscript with a function signature inspired by h2spec, [reaxe](https://github.com/jxnblk/reaxe), and other discussions including the one above. This is nowhere near being a component library -- just a dom node factory. I think Nanocomponent does a decent job for now. Will continue experimenting to try to find something workable. Closing for now. Thanks all for the input!
